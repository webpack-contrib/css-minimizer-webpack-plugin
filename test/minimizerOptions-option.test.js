import CopyPlugin from "copy-webpack-plugin";
import MiniCssExtractPlugin from "mini-css-extract-plugin";
import sugarss from "sugarss";

import CssMinimizerPlugin from "../src/index";

import { getCompiler, compile, readAsset } from "./helpers";

describe('when applied with "minimizerOptions" option', () => {
  it('matches snapshot for "discardComments" option (enable [default])', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/discardComments.css`,
      },
    });
    new CssMinimizerPlugin().apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "discardComments" option (disable)', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/discardComments.css`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        preset: ["default", { discardComments: false }],
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "discardComments" option (enable, with "removeAll" option)', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/discardComments.css`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        preset: ["default", { discardComments: { removeAll: true } }],
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "preset" option with require.resolve "String" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/order.css`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        preset: require.resolve("cssnano-preset-default"),
      },
    }).apply(compiler);

    const compiler2 = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/order.css`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        preset: require.resolve("cssnano-preset-default"),
      },
    }).apply(compiler2);

    const result1 = compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(
          `default-preset`,
        );
      }
    });

    const result2 = compile(compiler2).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler2, stats)).toMatchSnapshot(
          `preset-simple`,
        );
      }
    });

    return Promise.all([result1, result2]);
  });

  it('matches snapshot for "mergeRules" option (enable [default])', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/mergeRules.css`,
      },
    });
    new CssMinimizerPlugin().apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "mergeRules" option (disable)', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/mergeRules.css`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        preset: ["default", { mergeRules: false }],
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "discardEmpty" option (enable [default])', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/discardEmpty.css`,
      },
    });
    new CssMinimizerPlugin().apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "discardEmpty" option (disable)', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/minimizerOptions/discardEmpty.css`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        preset: ["default", { discardEmpty: false }],
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "parser" option with "Function" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/sugarss.js`,
      },
      module: {},
      plugins: [
        new CopyPlugin({
          patterns: [
            {
              context: `${__dirname}/fixtures/sss`,
              from: `index.sss`,
            },
          ],
        }),
        new MiniCssExtractPlugin({
          filename: "[name].css",
        }),
      ],
    });
    new CssMinimizerPlugin({
      test: /\.(css|sss)$/i,
      parallel: false,
      minimizerOptions: {
        processorOptions: {
          parser: sugarss,
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "parser" option with "String" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/sugarss.js`,
      },
      module: {},
      plugins: [
        new CopyPlugin({
          patterns: [
            {
              context: `${__dirname}/fixtures/sss`,
              from: `index.sss`,
            },
          ],
        }),
        new MiniCssExtractPlugin({
          filename: "[name].css",
        }),
      ],
    });
    new CssMinimizerPlugin({
      test: /\.(css|sss)$/i,
      minimizerOptions: {
        processorOptions: {
          parser: "sugarss",
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "stringifier" option with "Function" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/entry.js`,
      },
    });
    new CssMinimizerPlugin({
      parallel: false,
      minimizerOptions: {
        processorOptions: {
          stringifier: sugarss,
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "stringifier" option with "String" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/entry.js`,
      },
    });
    new CssMinimizerPlugin({
      minimizerOptions: {
        processorOptions: {
          stringifier: "sugarss",
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "syntax" option with "Function" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/sugarss.js`,
      },
      module: {},
      plugins: [
        new CopyPlugin({
          patterns: [
            {
              context: `${__dirname}/fixtures/sss`,
              from: `index.sss`,
            },
          ],
        }),
        new MiniCssExtractPlugin({
          filename: "[name].css",
        }),
      ],
    });
    new CssMinimizerPlugin({
      test: /\.(css|sss)$/i,
      parallel: false,
      minimizerOptions: {
        processorOptions: {
          syntax: sugarss,
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });

  it('matches snapshot for "syntax" option with "String" value', () => {
    const compiler = getCompiler({
      entry: {
        entry: `${__dirname}/fixtures/sugarss.js`,
      },
      module: {},
      plugins: [
        new CopyPlugin({
          patterns: [
            {
              context: `${__dirname}/fixtures/sss`,
              from: `index.sss`,
            },
          ],
        }),
        new MiniCssExtractPlugin({
          filename: "[name].css",
        }),
      ],
    });
    new CssMinimizerPlugin({
      test: /\.(css|sss)$/i,
      minimizerOptions: {
        processorOptions: {
          syntax: "sugarss",
        },
      },
    }).apply(compiler);

    return compile(compiler).then((stats) => {
      expect(stats.compilation.errors).toEqual([]);
      expect(stats.compilation.warnings).toEqual([]);

      for (const file in stats.compilation.assets) {
        // eslint-disable-next-line no-continue
        if (/\.js$/.test(file)) continue;
        expect(readAsset(file, compiler, stats)).toMatchSnapshot(file);
      }
    });
  });
});
