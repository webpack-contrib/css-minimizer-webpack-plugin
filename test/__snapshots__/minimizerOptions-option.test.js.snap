// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when applied with "minimizerOptions" option matches snapshot for "discardComments" option (disable): entry.css 1`] = `
"body{
  /* this is a comment */
  /*! this is a comment with exclamation mark */color:red}"
`;

exports[`when applied with "minimizerOptions" option matches snapshot for "discardComments" option (enable [default]): entry.css 1`] = `
"body{
  /*! this is a comment with exclamation mark */color:red}"
`;

exports[`when applied with "minimizerOptions" option matches snapshot for "discardComments" option (enable, with "removeAll" option): entry.css 1`] = `"body{color:red}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "discardEmpty" option (disable): entry.css 1`] = `"body{color:red}a{}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "discardEmpty" option (enable [default]): entry.css 1`] = `"body{color:red}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "mergeRules" option (disable): entry.css 1`] = `"body{color:red}body{font-weight:700}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "mergeRules" option (enable [default]): entry.css 1`] = `"body{color:red;font-weight:700}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "parser" option with "Function" value: index.sss 1`] = `"a{color:#000}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "parser" option with "String" value: index.sss 1`] = `"a{color:#000}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "preset" option with require.resolve "String" value: default-preset 1`] = `"p{border:0 solid var(--test);border-radius:var(--test);border-width:1px}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "preset" option with require.resolve "String" value: preset-simple 1`] = `"p{border:0 solid var(--test);border-radius:var(--test);border-width:1px}"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "stringifier" option with "Function" value: entry.css 1`] = `"body  color:reda  color:blue"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "stringifier" option with "String" value: entry.css 1`] = `"body  color:reda  color:blue"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "syntax" option with "Function" value: index.sss 1`] = `"a  color:#000"`;

exports[`when applied with "minimizerOptions" option matches snapshot for "syntax" option with "String" value: index.sss 1`] = `"a  color:#000"`;
